c8d/cache: Use the same cache logic as graphdrivers

vvoland opened the pull request:
**image/cache: Refactor backend specific code**

Move image store backend specific code out of the cache code and move it
to a separate interface to allow using the same cache code with
containerd image store.


**c8d/cache: Use the same cache logic as graphdrivers**

Implement the cache adaptor for containerd image store and use the same
cache logic.


**- How to verify it**
CI

**- Description for the changelog**


**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://github.com/moby/moby/assets/5046555/5a2bb79a-d0c5-4811-a545-769b6627ba39)

vvoland committed:
image/cache: Refactor backend specific code

Move image store backend specific code out of the cache code and move it
to a separate interface to allow using the same cache code with
containerd image store.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/cache: Use the same cache logic as graphdrivers

Implement the cache adaptor for containerd image store and use the same
cache logic.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Failed tests are expected Windows failures with containerd snapshotter.

laurazard commented on daemon/containerd/cache.go:
```diff
@@ -2,211 +2,238 @@ package containerd
+	return imgs
+}
+
+func findContentByUncompressedDigest(ctx context.Context, cs content.Manager, uncompressed digest.Digest) (content.Info, error) {
```
Sucks that we need to walk the content store for this 😓

vvoland replied:
Yeah, at least we have the label filter, so it's not as bad as just `Walk`, but still 😞

thaJeztah replied:
Silly question; this is to find compressed content based by the content pre compressing? What happens if there's multiple matches? (thinking here that images with different compression could have the same uncompressed content). Could that result in random results?

vvoland replied:
If the content store contains exactly the same layer but compressed with different compression algorithms (for example gzip and zstd) then yes, potentially the result would be random (it could either use gzip or zstd).

Considering this is a legacy builder, perhaps it would make sense to prefer gzip compressed blobs first?

thaJeztah replied:
`gzip` is the most portable, as it's required to be supported, so should always work.

In general, the _mixed compression_ scenario is the most problematic (some layers `gzip`, other layers `zstd`) because in that situation a client that doesn't support `zstd` can ignore half of the layers.

Of course there could be a situation where a user _wants to_ use `zstd`, so that could still be an option, but if it does, then _all layers_ should be `zstd`.

So I guess the "ideal" would be to make all layers match the same compression (if the first one was `zstd`, then prefer that).

That might be tricky though if the first match is "random", so preferring `gzip` would _probably_ be a good start (??); in that case the first layer would most likely be `gzip`, and follow-up layers as well (because of ordering).

Perhaps the only "full" solution would be having to do 2 passes;

- collect all possible options (with their compression)
- if there's mixed compression in some of them, then.. pick one that's available for all of them
- as a last resort, allow mixed compression

That's probably much too complicated though, but I'm wondering what BuildKit does in this scenario, or does it always recompress layers if needed?

vvoland replied:
Note that this only affects the `--cache-from` which already can result in picking up images that are not 100% matching.

Not sure how Buildkit handles it, I think `cache-from` behaves differently anyway, and even then  it has its own cache. With legacy builder we directly use images from the graph drivers and they're all stored in uncompressed form.

Considering that the legacy builder is deprecated anyway, I don't think it's worth to do anything more than just preferring to use the gzip.

thaJeztah replied:
> Note that this only affects the --cache-from which already can result in picking up images that are not 100% matching.

Good point  as well, also `--cache-from` with the classic builder has always been considered a "best effort" (but potentially insecure / false matches), so ... probably fine?

> Ah, forgot that the `content.Info` doesn't provide information about the content media type 🤦🏻 So we can't really determine if it's a gzip or not without reading it first.

Bummer!

Ah, I guess that makes sense if it follows the same approach as registry "blobs"? (i.e., the blob itself is just "data", and in that case the manifest defines how the data should be interpreted), or does it store that information, but just doesn't expose it?

vvoland replied:
No, it's not stored at all.
Although technically the content store writer interface allows to pass a descriptor: https://github.com/containerd/containerd/blob/90ec00c80cbdb4f2f031259a2415f519313fbef2/core/content/content.go#L188-L199

Also, `ReaderAt` accepts a descriptor, not just a digest, which hypothetically would mean that in future it _could_ take a media type into account and fail to return a reader if the media types doesn't match. But currently `MediaType` is ignored: https://github.com/containerd/containerd/blob/90ec00c80cbdb4f2f031259a2415f519313fbef2/core/content/proxy/content_store.go#L116-L129

thaJeztah replied:
Yeah, not sure if it can, because how to interpret content is a N:1 relation with the content itself. For example, the manifest-index below is perfectly valid; The referenced images manifests can be interpreted as either an OCI image or a "Docker" image; it's the same content but can be interpreted with different semantics. Similar situations could happen for other content;

```
docker buildx imagetools inspect FOOBAR
Name:      docker.io/library/FOOBAR
MediaType: application/vnd.oci.image.index.v1+json
Digest:    sha256:77906da86b60585ce12215807090eb327e7386c8fafb5402369e421f44eff17e

Manifests:
  Name:      docker.io/library/FOOBAR:latest@sha256:aa772c98400ef833586d1d517d3e8de670f7e712bf581ce6053165081773259d
  MediaType: application/vnd.oci.image.manifest.v1+json
  Platform:  linux/amd64

  Name:      docker.io/library/FOOBAR:latest@sha256:7409efd2c351d36aaca162069e56a19fa2633944215cc478832a72d7eadfaf10
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64
```

laurazard commented on daemon/containerd/cache.go:
```diff
@@ -2,211 +2,238 @@ package containerd
 			}
+
+			outImg.ContainerConfig = config
+			return errFound
```
Trying to understand this bit, are we doing this so we can stop as soon as we've found the image we want? Wonder if we should add a comment.

vvoland replied:
Yes, no need to read other images if we already found the one we're interested in. Can add a comment.

laurazard replied:
please 🙏

vvoland replied:
Let me know if you have some more comments, so I can batch them up in one force push 😄

vvoland replied:
Done, thanks!

laurazard commented on daemon/containerd/cache.go:
```diff
@@ -2,211 +2,238 @@ package containerd
-		for _, hi := range h {
-			if hi.ID != "<missing>" {
-				im, err := i.GetImage(ctx, hi.ID, backend.GetImageOpts{})
-				if err != nil {
-					return nil, err
-				}
-				images = append(images, im)
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return nil, err
 			}
+			log.G(ctx).Warnf("Could not look up %s for cache resolution, skipping: %+v", ref, err)
+			continue
 		}
+		cache.Populate(img)
 	}
 
-	return &imageCache{
-		lc: &localCache{
-			imageService: i,
-		},
-		images:       images,
-		imageService: i,
-	}, nil
+	return cache, nil
```
This ends up super similar to the graphdrivers one. Wonder if we can refactor this further.

vvoland replied:
Yeah it's exactly the same, just a different adaptor struct implementation.

Added a commit to extract the shared bits.

laurazard replied:
Nice! It looks great now :)

laurazard reviewed (approved):
LGTM overall! This is much better 🥳

vvoland committed:
imageService: Extract common code from `MakeImageCache`

Both containerd and graphdriver image service use the same code to
create the cache - they only supply their own `cacheAdaptor` struct.

Extract the shared code to `cache.New`.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

laurazard reviewed (approved):
Re-LGTM

thaJeztah reviewed (approved):
LGTM
